### ISSUE TITLE
Components not being unmounted when map is under <StrictMode>

### ISSUE THREAD

**Issue by urikphytech** [tomchentw/react-google-maps#921]
See question on SO along with codesandbox example of issue:
https://stackoverflow.com/questions/52755024/react-native-maps-doesnt-unmount-components-when-under-strictmode

**Comment by alexrussell** [comment]
Not normally one to :+1: issues, but being as this is new and has no comments yet, I didn't know if a "subscribe" would be good enough to show the intention, so just a quick comment to say I too am seeing this issue.

**Comment by alexrussell** [comment]
Actually my situation is a bit different to this one (but ultimately the same, see end of post). My issue is that under strict mode, sometimes markers appear on the map (if present on or around the "load" of the map) and then get re-rendered shortly after and then doubled up as the previous markers are not removed.

Also it's worth my saying that it's not just markers - it's possible all map adornments. Certainly it is happening for me for `Marker`s and a `Circle` too.

So, going back to my specific variant of this issue, I have been able to see what the cause of the issue is, though I'm not entirely sure of the best fix.

For me, it's because the Google Maps Marker object is set up and attached to the map in `construct` but only removed from the map in `componentWillUnmount`. This creates a situation in some cases whereby a marker that is constructed but never mounted leaves markers on the map.

So the underlying issue not actually a strict mode thing - simply an issue with the way that construct/mount/unmount is done in r-g-m.

However, strict mode has that exact situation happen: for me, without strict mode I never see a marker component construct without also being mounted, but under strict mode I see a construct with no mount. Why that is the case I am not sure, but there it is.

For reference, I was able to test this by simply adding some `console.log`s to the `constructor`, `cDM`, `cDU` and `cWU` methods of the r-g-m Marker class directly in my `node_modules`. You should be able to do the same, or use a proxy component as I do in a demo below.

So now, generalising this to the original issue encountered by @urikphytech, it appears to be the same thing - on mount, two markers are created but the first is constructed by never mounted. Here's a fork of @urikphytech's codesandbox demo with a proxy component (you'll want to open the console). Notice that as soon as you turn strict mode on, you get two "constructed" calls.

https://codesandbox.io/s/k95k7k1243

As such, I think the only fix for React's strict mode is to not do the Google Maps instantiation and initialisation in the component's constructor. Or maybe you can keep the initialisation in the constructor but only call `setMap` in `cDM` when you *know* you definitely mounted.

**Comment by alexrussell** [comment]
Also worth me saying that I added an answer to the SO post to the same effect but I think I did a semi-decent job of summarising what I said above, so [here's the link](https://stackoverflow.com/a/52851139/1551393) should @tomchentw or another maintainer wish to get the TL;DR version of the above.

**Comment by urikphytech** [comment]
Yes! Great job! As you said, there is actually a Marker being constructed but not mounted. 
This is due to:

> Strict mode can’t automatically detect side effects for you, but it can help you spot them by making them a little more deterministic. This is done by intentionally double-invoking the following methods:
> 
> Class component constructor method
> The render method
> setState updater functions (the first argument)
> The static getDerivedStateFromProps lifecycle

from: https://reactjs.org/docs/strict-mode.html

When moving the ```setMap``` call to ```cDM``` the codesandbox works. 

Another issue we're gonna have with migrating to v17 is that this library heavily uses the legacy context api. I'm planning on migrating it to the new context when I have the time...

Thanks!

**Comment by alexrussell** [comment]
Ahh yes I forgot about the fact that string mode intentionally messes you about in order to prod you in the direction of compliance. Well there you go, it's working! :)
